cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -305,14 +277,29 @@ func (s *service) Start(ctx context.Context, r *taskAPI.StartRequest) (*taskAPI.
 	if r.ExecID == "" {
 		cinit = container
 	} else {
-		s.pendingExecs[container]++
+		if _, initExited := s.containerInitExit[container]; initExited {
+			s.lifecycleMu.Unlock()
+			return nil, errdefs.ToGRPCf(errdefs.ErrFailedPrecondition, "container %s init process is not running", container.ID)
+		}
+		s.runningExecs[container]++
 	}
 	handleStarted, cleanup := s.preStart(cinit)
 	s.lifecycleMu.Unlock()
 	defer cleanup()
 
 	p, err := container.Start(ctx, r)
 	if err != nil {
+		// If we failed to even start the process, s.runningExecs
+		// won't get decremented in s.handleProcessExit. We still need
+		// to update it.
+		if r.ExecID != "" {
+			s.lifecycleMu.Lock()
+			s.runningExecs[container]--
+			if ch, ok := s.execCountSubscribers[container]; ok {
+				ch <- s.runningExecs[container]
```

Comment:
Seems like this could block if the for loop in the goroutine in `handleInitExit` has already received an event but hasn't run the defer.

Reply:
Hmm, the channel created in `s.handleInitExits` has a buffer of size `s.runningExecs[c]`, and we block new execs from being started before `handleInitExits` is called, so at most we'll process `s.runningExecs[c]` exits – so I don't think this would ever block.

Reply:
It seems feasible that the shim is not even notified that the container init exited until after we have gotten the failure to start the exec here but before we take the lock at L296?

Reply:
> It seems feasible that the shim is not even notified that the container init exited until after we have gotten the failure to start the exec here but before we take the lock at L296?

Following that thread:
- container init is started
- (1) container exec is started
	- we increment `s.runningExecs` in L284, then start the exec
- (1) the exec start call fails (L290)
- (2) container init exits and `s.processExits` acquires `s.lifecycleMu` and processes it
	- init exit gets stashed in `s.containerInitExits`, and `s.handleInitExit` gets called (and 2. unlocks `s.lifecycleMu`)
- (1) acquires the lock at L296, and decrements `s.runningExecs`
- (2)`s.handleInitExit` locks `s.lifecycleMu` checks `s.runningExecs`, finds 0, and immediately processes the init exit

Or:
- container init is started
- (1) container exec is started
	- we increment `s.runningExecs` in L284, then start the exec
- (1) the exec start call fails (L290)
- (2) container init exits and `s.processExits` acquires `s.lifecycleMu` and processes it
	- init exit gets stashed in `s.containerInitExits`, and `s.handleInitExit` gets called (and 2. unlocks `s.lifecycleMu`)
- (2)`s.handleInitExit` locks `s.lifecycleMu` checks `s.runningExecs`, finds 1, creates a buffered channel with length 1, launches the `goroutine` and unlocks `s.lifecycleMu`
- (1) acquires the lock at L296, and decrements `s.runningExecs`, and sends (a 0) on the channel
- (2) the goroutine receives on the channel, finds a 0, and processes the init exit.